enhancements/machine-config/manage-boot-images.md

```diff
@@ -0,0 +1,264 @@
+- `io.openshift.mco-managed-factory-image=` storing the original provider image reference prior to when the feature was turned on
+- `io.openshift.mco-managed-factory-secret=` storing the original stub secret name
+- `io.openshift.mco-managed-last-image=` storing the last provider image reference before an update
+- `io.openshift.mco-managed-last-secret=` storing the last stub secret name before an update
+
+##### via ConfigMap
+
+The MSC will maintain a backup of the following per machinset, with these keys:
+- `$(machine_set_name)-factory-image` storing the original provider image reference prior to when the feature was turned on
+- `$(machine_set_name)-factory-secret` storing the original stub secret name prior to the feature was turned on
+- `$(machine_set_name)-last-image` storing the last provider image reference before an update
+- `$(machine_set_name)-last-secret` storing the last stub secret name before an update
+
+##### Via new MachineSet fields
+
+Another proposed strategy is to add the required fields into the `.Status` section of a `MachineSet` object. This would be an API change.
+
+##### Mechanism
+
+A revert can be done by opting out the `MachineSet`, this will trigger the MSC to restore the MachineSet to last known good values stored from the backups. This is an important mitigation in case things go wrong(invalid bootimage references, incorrect patching... etc).
+
+The reason for keeping a factory version is in case the admin wishes to restore to factory values manually(they will have to opt-out of the feature first). It may also may aid in debugging.
+
+### Workflow Description
+
+It is important to note that there would be two "opt-in" knobs while this feature is under TechPreview. The admin would first have to turn on the feature gate, and then the opt-in mechanism. The secondary knob is necessary as some customers may want to keep their boot images static when this feature leaves TechPreview.
+Couple of strategies as before:
+
+##### via Annotations
+
+- To enroll a `MachineSet` for boot image updates, the cluster admin should add an annotation `io.openshift.mco-managed=true` to the `MachineSet`.
+- To un-enroll(and effectively revert) the `MachineSet` from boot image updates, the cluster admin should remove the `io.openshift.mco-managed=true` annotation from the `MachineSet`.
+
+##### via an Operator type 
+
+This is an API change to and would involve adding two new fields in the [operator types](https://github.com/openshift/api/blob/master/operator/v1/types_machineconfiguration.go) for the MCO:
+- `BootImageUpdateMode` This is an enum which can have three values: `Enabled`, `Selected` or `Disabled`
+- `BootImageUpdateEnrolledMachineSets` This is a list of enrolleds machinesets. When the above type is in the `Selected` mode, all machinesets in the list would be considered enrolled for updates.
+
+Here is a [rough PR](https://github.com/openshift/api/pull/1672) of what these API changes would look like.
```

Comment:
I've left some review on the API, once we have that updated, we should come back here and update here as well

Reply:
Makes sense, once we have consensus on the API I'll update this.

Reply:
I updated this to reflect the latest from our discussions.

Reply:
Updated to reflect the final shape.